File: testing-section/Classification-testing/part13-random-forest-classification/rfc.py
# Get the dataset
import pandas as pd # Data manipulation package
dataset = pd.read_csv('/home/felipe/Documentos/Machine Learning A-Z/Part 3 - Classification/' + 
	'Section 20 - Random Forest Classification/Random_Forest_Classification/Social_Network_Ads.csv')

# Check for missing values
# No missing values

# Check for categotical/hierarchical parameters
# No categorical neither hierarchical parameters, but the first two
# features are futile for this model.
dataset = dataset.iloc[:, 2:dataset.shape[1]]

# Split the dataset into independent matrix and 
# dependent array (python's way of machine learning)
ind_param = dataset.iloc[:, :-1].values
dep_param = dataset.iloc[:,  -1].values

# Feature scaling
# Random forest is not a distance-based algorithm, so it does not
# demand feature scaling.
from sklearn.preprocessing import StandardScaler as skl_ss
ind_param = skl_ss().fit_transform(ind_param.astype(float))

# Split the dataset into train and test sets
from sklearn.model_selection import train_test_split as skl_tts
ind_train_set, ind_test_set, dep_train_set, dep_test_set = skl_tts(ind_param, dep_param, test_size = 0.25)

# Fit the model
from sklearn.ensemble import RandomForestClassifier as skl_rfc
classifier = skl_rfc(
	n_estimators = 1000,
	criterion = 'entropy')
classifier.fit(ind_train_set, dep_train_set)

# Predict the test set
predictions = classifier.predict(ind_test_set)

# Confusion matrix
from sklearn.metrics import confusion_matrix as skl_cm
skl_cm(y_true = dep_test_set, y_pred = predictions)

# Plot the results
# =============================================
# Disclaimer: All the plotting code below is not mine.
# Font: Machine Learning A-Z™: Hands-On Python & R In Data Science - kernel_svm.py source code 
# =============================================
# Caution: Don't run that with a large (e.g 1000) number of trees/estimators without a powerful machine.
from matplotlib.colors import ListedColormap
import matplotlib.pyplot as plt 
import numpy as np

X_set, y_set = ind_train_set, dep_train_set

X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))

plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))

plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())

for i, j in enumerate(np.unique(y_set)):
	plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1], c = ListedColormap(('red', 'green'))(i), label = j)

plt.title('Kernel SVM (Training set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()
# ============================================

File: testing-section/Classification-testing/part13-random-forest-classification/rfc.R
# Get the dataset
dataset <- read.csv('/home/felipe/Documentos/Machine Learning A-Z/Part 3 - Classification/Section 14 - Logistic Regression/Social_Network_Ads.csv')

# Check for categorical/hierarchical parameters
dataset <- dataset[3:ncol(dataset)]
dataset$Purchased <- as.factor(dataset$Purchased) # Just to be sure that the model will work correctly

# Feature scaling? Just for plotting. Random Forest is not distance-based, so fs does not affect it.
dataset[, -ncol(dataset)] <- scale(apply(dataset[, -ncol(dataset)], 2, as.numeric))

# Split the dataset into test set and train set
library(caTools)
datasplit <- sample.split(dataset$Purchased, SplitRatio = 0.75)
train_set <- subset(dataset, datasplit)
test_set <- subset(dataset, !datasplit)

# Fit the model
library(randomForest)
set.seed(1234)
classifier <- randomForest(
 	formula = Purchased ~.,
 	ntree = 100,
 	data = train_set,
 	importance = TRUE)

classifier <- randomForest(
	formula = Purchased ~.,
	ntree = 100,
	x = train_set[-ncol(train_set)],
	y = train_set$Purchased,
	importance = TRUE)

# Predict the test set
prediction <- predict(object = classifier, newdata = test_set)

# Confusion matrix
table(test_set$Purchased, prediction)

which(test_set$Purchased != prediction) # w/ x and y: 54 69 73 76 77 91 93 96
										# w/ data 	: 54 69 73 76 77 91 93
										# what's that?!

# =============================================
# Disclaimer: All the plotting code below is not mine.
# Font: Machine Learning A-Z™: Hands-On Python & R In Data Science - kernel_svm.R source code 
# =============================================
# Caution: use this only with feature scaling.
library(ElemStatLearn)
set = train_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = predict(classifier, newdata = grid_set, type = 'class')
plot(set[, -3],
     main = 'Random Forest (Training set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))